### ISSUE TITLE
Bad performance when listing large directories

### ISSUE THREAD

**Issue by flewellyn** [GoogleCloudPlatform/gcsfuse#250]
I have a bucket which has a nested filesystem, containing satellite image files, often thousands per directory.  I find that when the number of files grows beyond a few thousand, performance for even a simple ls (without any options) becomes unacceptably slow.  

What's worse, even access to a single file, path fully specified, is slowed down by the directory listing, because gcsfuse seems to be doing multiple ListObjects calls per access to the single file.  

The upshot of this is that, if a given file is in its own subdirectory, rendering tiles from the file (using UMN Mapserver) is blazingly fast, taking less than a second per tile, but when read from a large directory, render time slows down to over 20 seconds.

Are there any options I can use to tune how this listing is done, or to cache the listing?

**Comment by jacobsa** [comment]
* Is it possible that you're using `--implicit-directories`? As [documented](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/docs/semantics.md#implicit-directories), this is slow because every name lookup requires a `ListObjects` request.

* A simple `ls` often involves a lot more than just reading the directory. On some Linux distributions for example, it `stat`s every entry in the directory, so this turns into thousands of `StatObject` calls to GCS. You can confirm this with `strace`, and often turn it off with `ls -1`.

See #234 for a similar issue. If you still have concers, please collect a log using `--debug_fuse` and `--debug_gcs`.

**Comment by flewellyn** [comment]
Thanks for your quick response.  I am not using the --implicit-directories option, because of its documented performance impact.  Also, it isn't necessary, since in this setup, the bucket will have multiple readers, but only one writer.

I found that disabling the ls color option that Ubuntu set to "auto" by default caused ls to be very much faster: a debug trace of GCSFuse showed no StatObject calls.  What puzzled me was that it still did StatObject calls when I tried a wildcard listing, a la "ls Foo*".  

I should note that I did have a look at issue #234 , and tried using a modified GCSFuse, with the indicated code changed to send MaxResults of 1000.  Didn't seem to make much difference.

What's puzzling me is that reading from a single file, where there is no directory listing needed, is still hitting ListObjects so hard, when the file path is already known.  I enabled stat and type caching, but this directory entry doesn't seem to be preserved in any cache.

**Comment by jacobsa** [comment]
Everything except the last paragraph sounds expected -- if you are asking the file system to stat the objects, then you'll need to send `StatObject` calls. But the last paragraph is a surprise. Please post debug logs.

**Comment by flewellyn** [comment]
I was gathering the debug logs, when I noticed in them the source of the problem.  It has to do with the GIS library I am using, GDAL, which tries to open a number of common "subsidiary" files whenever it opens a file as a datasource.  Each time it tried to find one of these auxiliary files, it had to do another directory listing.  I had to configure it not to do that, at which point suddenly the performance problems disappeared.

Sorry to be a bother!  Thanks for your help.
